### DISCUSSION TITLE
How to use `kube::Api::create` with raw string YAMLs?

### DISCUSSION THREAD

**Discussion by mati865** [kube-rs/kube#1564] [Q&A] [answered]
According to the docs this function accepts "Raw string YAML" but I cannot find the obvious way to use it.

My code (minimized):
```rust
fn foo(namespace: &str, raw_template: String) -> anyhow::Result<()> {
        let k8s_client = Client::try_default().await?;
        let api = Api::namespaced(k8s_client, &namespace);
        api.create(&PostParams::default(), &raw_template)
            .await?;
}
```
shows an error:
```
error[E0277]: the trait bound `std::string::String: k8s_openapi::Metadata` is not satisfied
  --> foo/src/bar.rs:40:19
   |
40 |         let api = Api::namespaced(k8s_client, &namespace);
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `k8s_openapi::Metadata` is not implemented for `std::string::String`, which is required by `std::string::String: kube::Resource`
   |
   = help: the following other types implement trait `k8s_openapi::Metadata`:
             APIService
             Binding
             CSIDriver
             CSINode
             CSIStorageCapacity
             CertificateSigningRequest
             ClusterRole
             ClusterRoleBinding
           and 77 others
   = note: required for `std::string::String` to implement `kube::Resource`
```

My use case is trying to deploy YAMLs that contain multiple resources separated by `---`.

**Comment by clux** [comment, marked as answer]
if the yaml are arbitrary resources, then you'd probably want to use the dynamic api.

Take a look at the kubectl example in the apply section: https://github.com/kube-rs/kube/blob/7ff120a1cad8797b098919eada1485573ea722c5/examples/kubectl.rs#L156-L181

i suspect that's pretty similar to what you want (it's using server-side apply - via api.patch rather than api.create, but this should be better than using create anyway)

**Reply by mati865** [reply]
Thank you for the reply, that's great solution.

At first I planned to use server-side, but since I'm doing only a proof of concept I decided to start with `create` as according to the docs I could just call it with a yaml. If that's not possible maybe docs need a correction?

**Reply by clux** [reply]
you should be able to do basically the same as above there with `api.create` without much trouble (s/PatchParams/PostParams ++).

i just try to steer people away from the older (non-idempotent) `create`/`replace` interfaces unless there's a particular need need to use them (because they come with potential needs for conflict resolution). you should be able to call either with a full object, `.patch` is just a little more lenient in accepting partial objects / patches.
